### PR TITLE
Update and Refactor `find_MAP` and `fit_laplace`

### PR DESCRIPTION
**Pull request by jessegrabowski**
I just made an update to better-optimize that uses hessian matrix caching for better performance. This is something we can immediately take advantage of with `find_MAP`, re-using sub-computation from the loss or gradient in the hessian computation. This PR updates the functions generated by find_MAP to take advantage of this.

While I was at it, I went ahead and did some cleanup and reorganization of the code. In particular:

1. I split up the files into smaller, more logical groupings. I moved all the files to a `laplace_approx` submodule.
2. `find_MAP` now returns an idata. This is more consistent with all the other PyMC sampling function -- it's weird to get back a dictionary in this one case.
3. When available, `find_MAP` will now always store the inverse hessian. This is done to try to avoid an extra function compilation when it is used in conjunction with `fit_laplace`.
4. `fit_laplace` was a really dumb function that was inexplicably sampling from scipy distributions. This required a ton of unnecessary work. If only we had a PPL that could help sample from complicated distributions...

`fit_laplace` still isn't perfect. I wanted to store both the value variables and the transformed RVs as deterministics in a pymc model and sample them directly, but that doesn't appear to work -- maybe this is a bug? I ended up doing two passes, once for the constrained RVs, then a second pass for the unconstrained. It would be good to minimize that.

I also removed as many little options that were floating around as possible. These function signatures were already horrible.

Finally, I eliminated a lot of test parameterizations to speed the CI up, but also added a lot of new tests for functions that were previously not covered. Hopefully it's still net positive.

### CODE REVIEW

File: pymc_extras/inference/laplace_approx/laplace.py [hunk 1]
```diff
@@ -147,189 +139,84 @@ def get_conditional_gaussian_approximation(
...
-        if on_bad_cov == "error":
-            raise np.linalg.LinAlgError(
-                "Inverse Hessian not positive-semi definite at the provided point"
+        cursor = 0
+        for rv, info in zip(free_rvs, raveled_vars.point_map_info):
+            transform = model.rvs_to_transforms[rv]
+            name, shape, size, dtype = info
+            unconstrained_value = (
+                laplace_approximation[..., cursor : cursor + size]
+                .reshape((chains, draws, *shape))
```
Comment (ricardoV94) [inline pymc_extras/inference/laplace_approx/laplace.py:174]: Can you add a test with a variable that has different dimensionality in the unconstrained space (like a dirichlet)?

File: pymc_extras/inference/laplace_approx/laplace.py [hunk 2]
```diff
@@ -147,189 +139,84 @@ def get_conditional_gaussian_approximation(
...
+
+            pm.Deterministic(
+                rv.name,
+                (
+                    transform.backward(unconstrained_value, *rv.owner.inputs)
+                    if transform is not None
+                    else unconstrained_value
+                ),
+                dims=dims,
             )
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pymc_extras/inference/laplace_approx/laplace.py]: you have to look at what model.replace_rvs_by_values does. If you check the old approach that used `model.unobserved_value_vars` you'll see it uses that under the hood.

This is only a problem because you're using the unconstructed variables from fitting. If you use the constrained ones transforms and values don't show up.

[the remark this answers is not in the document]
Comment (jessegrabowski) [reply pymc_extras/inference/laplace_approx/laplace.py]: I am enamored with the idea of manipulating pymc models directly, rather than rooting around in the backend. I think it leads to conceptually cleaner code.

What I didn't like about the old implementation was that I was drawing from a statsmodels MvN then passing those draws into a compiled function. If I go back, I'd rather do that all symbolicaly and pass the mean and covariance as the function inputs.

[the remark this answers is not in the document]
Comment (ricardoV94) [reply pymc_extras/inference/laplace_approx/laplace.py]: Just do a graph replace on the joined tensor. The thing is you're not working with the forward model only because you have value transforms in the way. Also the model is just a fancy dict for pytensor variables so working directly with those seems fine(r) than trying to force things into a model

File: pymc_extras/inference/laplace_approx/laplace.py (hunk 1 above)
Comment (jessegrabowski) [reply pymc_extras/inference/laplace_approx/laplace.py]: Nevermind figured it out

**Comment by jessegrabowski** [comment]
This should be ready to go. Last changes:

- `find_laplace` should work in all cases, including when the value variables have a different shape as the RVs (added test for this). I brought back the old way of making the batched RVs, then combine it with the new pymc model approach to make the raveled vector. I think the result is quite nice.

- `find_MAP` posterior has chain and draw, to stick fast to the arviz convention (but it can be easily squeezed away by the user)

- I found some unused helpers and removed them. Also removed the `utilities` file, because I'm sick of those and I never remember where I put anything.

File: pymc_extras/inference/laplace_approx/scipy_interface.py [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    orig_loss_fn = f_fused.vm.jit_fn
+
+    if use_hess:
+
+        @jax.jit
+        def loss_fn_fused(x):
+            loss_and_grad = jax.value_and_grad(lambda x: orig_loss_fn(x)[0])(x)
+            hess = jax.hessian(lambda x: orig_loss_fn(x)[0])(x)
+            return *loss_and_grad, hess
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pymc_extras/inference/laplace_approx/scipy_interface.py:53]: Note: You didn't update the fix in the original commit where it's introduced

File: pymc_extras/inference/laplace_approx/laplace.py [hunk 4]
```diff
@@ -147,189 +146,110 @@ def get_conditional_gaussian_approximation(
...
-            dtype=unconstrained_vector.type.dtype,
-        )
-        batched_rvs = pytensor.graph.vectorize_graph(
-            constrained_rvs, replace={unconstrained_vector: batched_values}
-        )
+    coords = model.coords | {"chain": range(chains), "draw": range(draws)}
 
-        f_constrain = pm.compile(inputs=[batched_values], outputs=batched_rvs, **compile_kwargs)
-        posterior_draws = f_constrain(posterior_draws)
+    # There are corner cases where the value_vars will not have the same dimensions as the random variable (e.g.
```
[the remark this answers is not in the document]
Comment (ricardoV94) [reply pymc_extras/inference/laplace_approx/laplace.py:227]: But we don't have coords for those ofc

File: pymc_extras/inference/laplace_approx/laplace.py [hunk 5]
```diff
@@ -147,189 +146,110 @@ def get_conditional_gaussian_approximation(
...
-    H_inv: np.ndarray,
-    model: pm.Model | None = None,
-    chains: int = 2,
-    draws: int = 500,
-    transform_samples: bool = False,
-    progressbar: bool = True,
-    random_seed: int | np.random.Generator | None = None,
-    compile_kwargs: dict | None = None,
-) -> az.InferenceData:
+def unstack_laplace_draws(idata, model):
```
[the remark this answers is not in the document]
Comment (jessegrabowski) [reply pymc_extras/inference/laplace_approx/laplace.py:207]: Let me double-check, might be a holdover from the old stuff that I got confused about.

[the remark this answers is not in the document]
Comment (jessegrabowski) [reply pymc_extras/inference/laplace_approx/laplace.py:217]: sure, but that's hard to work with, so why not unravel it into variables for the user?

[the remark this answers is not in the document]
Comment (ricardoV94) [reply pymc_extras/inference/laplace_approx/laplace.py:217]: Do you really care about the unconstrained simplex draws? Aren't the constrained ones enough? Users aren't usually asking for it in regular mcmc?

[the remark this answers is not in the document]
Comment (ricardoV94) [reply pymc_extras/inference/laplace_approx/laplace.py:217]: also this doesn't seem Laplace specific, after all nuts also samples an unconstrained raveled vector and we don't go out of our way to offer those draws split to the user.

We could have a general helper like the one proposed here: https://github.com/pymc-devs/pymc/issues/6721

Which is not tied to a specific sampling or fitting method

[the remark this answers is not in the document]
Comment (jessegrabowski) [reply pymc_extras/inference/laplace_approx/laplace.py:217]: I agree it's not sampler specific, except to the extent that laplace isn't really a sampling method, it's a multivariate normal that I'm pretending is a sampling method. I'm just trying to be supremely conservative.

I've been thinking about what nutpie does, returning both transformed and untransformed in the posterior. I think it's correct to return the unconstrained values, since these are the "true" approximation. But I really, really hate that nutpie shoves it all into the posterior group -- it creates a ton of plotting clutter. I thought adding the unconstrained_posterior group in the return here is a nice middle ground.

[the remark this answers is not in the document]
Comment (ricardoV94) [reply pymc_extras/inference/laplace_approx/laplace.py:217]: On the auto dims part I'm not sure it's sound to try to infer dims/coords. You could have a weird transform that swaps two dims or whatever. Just because something has the same length and comes on the same axis doesn't mean it has the same dims (although it practically always has).

With the dims module otoh you wouldn't have to guess because we know the dims of the value variables and their transforms. So you may be trying to reinvent a wheel here without a need?

[the remark this answers is not in the document]
Comment (jessegrabowski) [reply pymc_extras/inference/laplace_approx/laplace.py:217]: Leaving away dims is an absolutely miserable user experience. I'd rather get that transform case wrong (or error out) than just live up and leave 99% of cases unlabeled.

[the remark this answers is not in the document]
Comment (ricardoV94) [reply pymc_extras/inference/laplace_approx/laplace.py:217]: Okay. Still seems a niche place to invent this stuff here

[the remark this answers is not in the document]
Comment (ricardoV94) [reply pymc_extras/inference/laplace_approx/laplace.py:217]: yeah

[the remark this answers is not in the document]
Comment (jessegrabowski) [reply pymc_extras/inference/laplace_approx/laplace.py:217]: `pm.sample_posterior_predictive` filters out value variable names [here](https://github.com/pymc-devs/pymc/blob/ae43026767209d4a78ce812e08b20210dacb8442/pymc/sampling/forward.py#L890), so doing everything symbolically in one shot doesn't work.

I'm sticking with what I have for now, I really just need this PR to be done.

**Review by ricardoV94 — APPROVED** [review]
